File: 3. Stacks and Queues/Stack.cpp
#include "Stack.h"
#include <iostream>
using std::cout;
using std::endl;

Stack::Stack() {
	top = NULL;
	size = 0;
}

int Stack::pop() {
	if(top == NULL) {
		// cout << "Not found" << endl;
		return 404;
	} else {
		int result = top->getdata();
		top = top->getnext();
		size--;
		return result;
	}
}

void Stack::push(int value) {
	StackNode* newnode = new StackNode(value);
	newnode->setnext(top);
	top = newnode;
	size++;
}

int Stack::peek() const {
	if(top == NULL) {
		// cout << "Nothing on Stack" << endl;
		return 404;
	} else {
		return top->getdata();
	}
}

bool Stack::isEmpty() const {
	if(top == NULL) {
		return 1;
	} else {
		return 0;
	}
}

void Stack::print() {
	if(top == NULL) {
		cout << "Nothing on Stack" << endl;
	} else {
		StackNode* temp = top;
		while(temp != NULL) {
			cout << temp->getdata();
			if(temp->getnext() != NULL) {
				cout << "->";
			}
			temp = temp->getnext();
		} 
	}
}


int Stack::getSize() const {
	return size;
}

StackNode* Stack::getTop() const {
	return top;
}

File: c_stackqueue/Stack.java
package c_stackqueue;

/*
	
	Interface of Stack
	pop(): Remove the top item from the stack
	push(item): Add an item to the top of the stack
	peek(): return the top of the stack
	isEmpty(): return true if and only if the stack is empty
	

*/

public class MyStack {

	private static class StackNode {
		int data;
		StackNode next;

		public StackNode(int data) {
			this.data = data;
		}
	}

	private StackNode top;

	public int pop() {
		if(top == null) throw new EmptyStackException();
		int topval = top.data;
		top = top.next;
		return topval;
	}

	public void push(int data) {
		StackNode newNode = new StackNode(data);
		newNode.next = top;
		top = newNode;
	}

	public int peek() {
		if(top == null) throw new EmptyStackException();
		return top.data;
	}

	public boolean isEmpty() {
		return top==null;
	}
}